### PR TITLE
chore: refactoring show preview classes

### PR DESCRIPTION
**Pull request by jonathanyeung**
## chore: refactoring show preview classes

This change refactors how the plugin-side classes around Preview work. The primary motivation is to get rid of circular dependencies in this area.
- There's now a unified way for plugin code to perform Preview operations - by getting an instance of `PreviewProxy`. This interface now completely hides the `vscode.webviewPanel`, so no code calling preview has to worry about preview lifecycle events anymore. This interface also now exposes methods `isOpen` and `isVisible`.  If more properties are needed later, we can add them. Using this interface should help us avoid introducing circular dependencies in this area going forward.
- All event registration around preview (like changing editors, updating text document) are encapsulated within `PreviewPanel`. Construction of PreviewPanel is abstracted through `PreviewPanelFactory`, which returns a `PreviewProxy` (_not_ a PreviewPanel). This means we can pull out all lifecycle events outside of the window/workspace watchers (and hence remove more circular dependencies). The one exception is with the `automaticallyShowPreview` feature, which still needs to listen to onEditorChanged outside of `PreviewPanel`, as it needs to potentially activate the panel if it's not currently opened.
- For simplification, I've pulled out the methods surrounding preview-link-click-handling logic into a separate interface and class in `PreviewLinkHandler.ts`. This also helps with testing.
- Most logic in `ShowPreviewCommand` have been pulled out and moved to `PreviewPanel` and `PreviewLinkHandler`

#### Manual Tests:

There's a high chance of regression with this change, so it'd be good to have some eyes on it via Nightly. I've done some manual testing, and so far the following scenarios look good:

- Regular Preview Operations
- Enabled automaticallyShowPreview - feature worked; no undesired refocus effect when having the preview in the same editor group as a note.
- Opened a markdown file not part of Dendron workspace via context menu; preview opened fine.

####  Existing Preview Issues

I did not actively try to fix existing preview bugs with this change (which seem to be happening on the webview side), those are out of scope for this particular change and include:
- preview no longer updating after a while
- preview going blank after a while
- metadata info (such as backlinks and tags) disappearing once you start typing.

---

# Pull Request Checklist

If this is your first time submitting a pull request to Dendron, copy and paste the full [Dendron Review Checklist](https://docs.dendron.so/notes/1EoNIXzgmhgagqcAo9nDn.html) into this request and check off each item as necessary. 

This template contains the short checklist which is used by the Dendron core team. 

## Testing
- [ ] [Write Tests](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#writing-tests) 
- [ ] [Confirm existing tests pass](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#executing-tests)
- [ ] [Confirm manual testing](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#manual-testing) 
- [ ] If your tests changes an existing snapshot, make sure that snapshots are [updated](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#updating-test-snapshots)
- [ ] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), make sure that it is included in the [test workspace](https://docs.dendron.so/notes/dtMsF12SF2SUhLN10sYe2.html)
### Docs
- [ ] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
### Analytics
- [ ] if you are adding analytics related changes, make sure the [Telemetry](https://docs.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### CODE REVIEW

File: packages/plugin-core/src/components/views/PreviewPanel.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+  /**
+   * Show the preview.
+   * @param note - if specified, this will override the preview contents with
+   * the contents specified in this parameter. Otherwise, the contents of the
+   * preview will follow default behavior (it will show the currently in-focus
+   * Dendron note).
+   */
+  show(note?: NoteProps): void {
+    if (this._panel) {
```
Comment (tma66) [inline packages/plugin-core/src/components/views/PreviewPanel.ts:65]: Nit, but you can reuse your methods you defined. `if (this.isOpen() && !this.isVisible())`

**Review by tma66 — COMMENTED** [review]
This looks pretty good! I focused mostly on the `automaticallyShowPreview` logic. Other than couple minor things, there's one behaviour that differs from existing one (one upon startup)

**Comment by SeriousBug** [comment]
Other than the issue with preview always launching at startup regardless of the setting, looks good to me.

**Review by kevinslin — CHANGES_REQUESTED** [review]
nice job. one request - can you update docs in [[Views|dendron://dendron.docs/pkg.plugin-core.dev.cook#views]] so folks know how to access the preview going forward?

also, for major refactoring operations, can you also add it to `dev.process.close-loop` about announcing this in the #dev channel when merged? similar to https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330

File: packages/plugin-core/src/components/views/PreviewProxy.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  /**
+   * Is the view visible?
+   */
+  isVisible(): boolean;
+
+  /**
+   * Is the view open? A view can be opened but not visible - for example, if a
+   * tab for it exists but it's not in focus in the editor group. If a view is
+   * open, then it is necessarily visible.
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/plugin-core/src/components/views/PreviewProxy.ts:32]: good catch!

File: packages/plugin-core/src/components/views/PreviewPanel.ts (hunk 1 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/components/views/PreviewPanel.ts:65]: made some slight modifications here - but one Typescript thing - if it's not explicit that `this._panel` is not undefined via the if condition, then its usage on line 67 will result in a 'object possibly undefined' warning.

File: packages/plugin-core/src/_extension.ts [hunk 3]
```diff
@@ -777,7 +777,7 @@ export async function _activate(
       // If automaticallyShowPreview = true, display preview panel on start up
       const note = WSUtils.getActiveNote();
       if (note) {
-        PreviewPanelFactory.getProxy(getExtension()).showPreviewAndUpdate(note);
+        PreviewPanelFactory.create(getExtension()).show(note);
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/plugin-core/src/_extension.ts:780]: good catch!

**Comment by jonathanyeung** [comment]
> nice job. one request - can you update docs in [[Views|dendron://dendron.docs/pkg.plugin-core.dev.cook#views]] so folks know how to access the preview going forward?
> 
> also, for major refactoring operations, can you also add it to `dev.process.close-loop` about announcing this in the #dev channel when merged? similar to [#2188 (review)](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

Updated docs at the following places:
- [[Views|dendron://dendron.docs/dev.cook#views]]
- [[Close Loop|dendron://dendron.docs/dev.process.close-loop]]

**Review by kevinslin — APPROVED** [review]

**Pull request merged**